system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+class OpenFileDescriptor {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  OpenFileDescriptor(uint32_t offset, std::shared_ptr<File> file);
+
+  std::shared_ptr<File>& getFile();
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<OpenFileDescriptor>> entries;
+  std::mutex mutex;
+
+  FileTable() {
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdinFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdoutFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StderrFile>()));
+  }
+
+public:
+  // Handle represents an RAII wrapper object. Access to the FileTable must go through a Handle.
+  // A Handle holds FileTable's lock for the duration of its lifetime.
+  class Handle {
+    FileTable& fileTable;
```

Comment:
Yes, but we make it a local static variable so it will be lazily initialized. That means we don't have a name to refer to it with elsewhere, so we need to store a reference to it.

Reply:
Yeah, I can make the comment clearer here for sure